File: othello_csharp/Othello/Board.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;

namespace Othello
{
    /// Handles game board state and logic.
    internal class Board
    {
        private readonly List<Disk> _board;
        private readonly List<Square> _emptySquares;
        private readonly List<int> _indices;
        private readonly int _size;

        private static readonly int[][] StepDirections =
        {
            new [] {-1,-1},
            new [] {-1, 0},
            new [] {-1, 1},
            new [] { 0,-1},
            new [] { 0, 1},
            new [] { 1,-1},
            new [] { 1, 0},
            new [] { 1, 1}
        };

        public Board(int size)
        {
            _size = size;
            var numSquares = _size * _size;
            // init game board with empty disks
            _board = Enumerable.Repeat(Disk.Empty, numSquares).ToList();

            // set starting positions
            var row = _size % 2 == 0 ? (_size - 1) / 2 : (_size - 1) / 2 - 1;
            var col = _size / 2;
            _board[row * _size + row] = Disk.Black;
            _board[row * _size + col] = Disk.White;
            _board[col * _size + row] = Disk.White;
            _board[col * _size + col] = Disk.Black;

            // index list (0...size) to avoid repeating same range in for loops
            _indices = Enumerable.Range(0, _size).ToList();

            // keep track of empty squares on board to avoid checking already filled positions
            _emptySquares = new List<Square>(numSquares);
            foreach (var square in _indices
                .SelectMany(y => _indices, (y, x) => new Square(x, y))
                .Where(square => GetSquare(square) == Disk.Empty))
            {
                _emptySquares.Add(square);
            }
        }

        /// Return true if board contains empty squares
        public bool CanPlay()
        {
            return _emptySquares.Any();
        }

        /// Update board for given disk placement.
        public void PlaceDisc(Move move)
        {
            var start = move.Square;
            if (GetSquare(start) != Disk.Empty)
            {
                throw new ArgumentException($"Trying to place disk to an occupied square {start}!");
            }
            SetSquare(start, move.Disk);
            _emptySquares.Remove(start);
            foreach (var dir in move.Directions)
            {
                var pos = start + dir;
                while (GetSquare(pos) == move.Disk.OtherDisk())
                {
                    SetSquare(pos, move.Disk);
                    pos += dir;
                }
            }
        }

        /// Returns a list of possible moves for given player.
        public List<Move> PossibleMoves(Disk color)
        {
            var moves = new List<Move>();
            var other = color.OtherDisk();
            foreach (var square in _emptySquares)
            {
                var value = 0;
                var directions = new List<Square>();
                foreach (var dir in StepDirections)
                {
                    var step = new Square(dir[0], dir[1]);
                    var pos = square + step;
                    // next square in this directions needs to be opponents disk
                    if (GetSquare(pos) != other)
                    {
                        continue;
                    }
                    var steps = 0;
                    // keep stepping forward while opponents disks are found
                    while (GetSquare(pos) == other)
                    {
                        ++steps;
                        pos += step;
                    }
                    // valid move if a line of opponents disks ends in own disk
                    if (GetSquare(pos) != color)
                    {
                        continue;
                    }
                    value += steps;
                    directions.Add(step);
                }
                if (value > 0)
                {
                    moves.Add(new Move(square, value, color, directions));
                }
            }
            if (moves.Any())
            {
                moves.Sort(delegate (Move left, Move right)
                {
                    var value = right.Value.CompareTo(left.Value);
                    return value == 0 ? left.Square.X.CompareTo(right.Square.X) : value;
                });
            }
            return moves;
        }

        /// Print available move coordinates and resulting points gained.
        public void PrintMoves(IReadOnlyCollection<Move> moves)
        {
            ColorPrint.WriteLine($"  Possible moves ({moves.Count}):", Color.Yellow);
            // convert board from Disk enums to strings
            var boardStr = new List<string>(_board.Count);
            boardStr.AddRange(_board.Select(disk => disk.BoardChar()));
            foreach (var move in moves)
            {
                Console.WriteLine($"  {move}");
                var (x, y) = move.Square;
                boardStr[y * _size + x] = ColorPrint.Get(move.Value, Color.Yellow);
            }
            // print board with move positions
            Console.Write("   ");
            foreach (var i in _indices)
            {
                Console.Write($" {i}");
            }
            foreach (var y in _indices)
            {
                Console.Write($"\n  {y}");
                foreach (var x in _indices)
                {
                    Console.Write($" {boardStr[y * _size + x]}");
                }
            }
            Console.WriteLine("");
        }

        /// Print current score for both players.
        public void PrintScore()
        {
            var (black, white) = PlayerScores();
            Console.WriteLine("\n" + this);
            Console.Write($"Score: {ColorPrint.Get(black, Disk.Black.DiskColor())} | " +
                                 $"{ColorPrint.Get(white, Disk.White.DiskColor())}\n");
        }

        /// Calculates the final score and returns the winning player.
        public Disk Result()
        {
            var sum = Score();
            if (sum == 0)
            {
                return Disk.Empty;
            }
            return sum > 0 ? Disk.White : Disk.Black;
        }

        /// Format game board to string
        public override string ToString()
        {
            var text = _indices.Aggregate(" ", (current, i) => current + $" {i}");
            foreach (var y in _indices)
            {
                text += $"\n{y}";
                text = _indices
                    .Select(x => _board[y * _size + x])
                    .Aggregate(text, (current, disk) => current + $" {disk.BoardChar()}");
            }
            return text;
        }

        /// Check that the given coordinates are inside the board
        private bool CheckCoordinates(int x, int y)
        {
            return 0 <= x && x < _size && 0 <= y && y < _size;
        }

        /// Count and return the number of black and white disks.
        private (int, int) PlayerScores()
        {
            var black = 0;
            var white = 0;
            foreach (var disk in _board)
            {
                switch (disk)
                {
                    case Disk.White:
                        ++white;
                        break;
                    case Disk.Black:
                        ++black;
                        break;
                    case Disk.Empty:
                        break;
                    default:
                        throw new ArgumentOutOfRangeException();
                }
            }
            return (black, white);
        }

        /// Returns the total score (positive means more white disks and negative means more black disks).
        private int Score()
        {
            return _board.Sum(x => Convert.ToInt32(x));
        }

        /// Sets the given square to given value.
        private void SetSquare(Square square, Disk disk)
        {
            var (x, y) = square;
            if (!CheckCoordinates(x, y))
            {
                throw new ArgumentException($"Invalid coordinates ({x},{y})!");
            }
            _board[y * _size + x] = disk;
        }

        /// Returns the state of the board (empty, white, black) at the given coordinates.
        private Disk? GetSquare(Square square)
        {
            var (x, y) = square;
            if (!CheckCoordinates(x, y))
            {
                return null;
            }
            return _board[y * _size + x];
        }
    }
}

File: othello_swift/Sources/othello_swift/board.swift

class Board {
    let size: Int
    var board: [Disk]
    var empty_squares = Set<Square>()
    let step_directions = [[0, -1], [0, 1], [1, 0], [-1, 0], [1, -1], [1, 1], [-1, 1], [-1, -1]]

    init(size: Int) {
        self.size = size
        // init game board with empty disks
        self.board = [Disk](repeating: Disk.empty, count: size * size)

        // set starting positions
        let pos = (self.size - 1) / 2
        let row: Int = self.size % 2 == 0 ? pos : pos - 1
        let col: Int = self.size / 2
        self.board[row * self.size + row] = Disk.black
        self.board[row * self.size + col] = Disk.white
        self.board[col * self.size + row] = Disk.white
        self.board[col * self.size + col] = Disk.black

        // keep track of empty squares on board to avoid checking already filled positions
        for y in 0 ..< self.size {
            for x in 0 ..< self.size {
                if self.board[y * self.size + x] == Disk.empty {
                    self.empty_squares.insert(Square(x, y))
                }
            }
        }
    }

    /// Return true if board contains empty squares.
    func can_play() -> Bool {
        !self.empty_squares.isEmpty
    }

    /// Update board for given disk placement.
    func place_disk(move: Move) {
        let start = move.square
        self.set_square(start, move.disk)
        self.empty_squares.remove(start)
        for step in move.directions {
            var pos = start + step
            while self.get_square(pos) == move.disk.other_disk() {
                self.set_square(pos, move.disk)
                pos += step
            }
        }
    }

    /// Returns a list of possible moves for given player.
    func possible_moves(disk: Disk) -> [Move] {
        var moves = [Move]()
        let other: Disk = disk.other_disk()
        for square in self.empty_squares {
            var value = 0
            var directions = [Square]()
            for dir in self.step_directions {
                let step = Square(dir[0], dir[1])
                var pos = square + step
                // next square in this directions needs to be opponents disk
                if self.get_square(pos) != other {
                    continue
                }
                var steps = 0
                // keep stepping forward while opponents disks are found
                while self.get_square(pos) == other {
                    steps += 1
                    pos += step
                }
                // valid move if a line of opponents disks ends in own disk
                if self.get_square(pos) != disk {
                    continue
                }
                value += steps
                directions.append(step)
            }
            if value > 0 {
                moves.append(Move(square, value, disk, directions))
            }
        }
        return moves.sorted(by: { $0.value > $1.value })
    }

    /// Print available move coordinates and resulting points gained.
    func print_possible_moves(_ moves: [Move]) {
        print("  Possible plays (\(moves.count)):".yellow())
        // convert board from Disk enums to strings
        var board_str = self.board.map { $0.board_char() }
        for move in moves {
            print("  \(move)")
            board_str[move.square.y * self.size + move.square.x] = "\(move.value)".yellow()
        }
        // print board with move positions
        print("   ", terminator: "")
        for x in 0 ..< self.size {
            print(" \(x)".dim(), terminator: "")
        }
        for y in 0 ..< self.size {
            print("\n  \(y)".dim(), terminator: "")
            for x in 0 ..< self.size {
                print(" \(board_str[y * self.size + x])", terminator: "")
            }
        }
        print("")
    }

    /// Returns the winner color.
    func result() -> Disk {
        let sum: Int = self.score()
        switch sum {
            case ..<0:
                return Disk.white
            case 1...:
                return Disk.black
            default:
                return Disk.empty
        }
    }

    /// Print current score for both players.
    func print_score() {
        let score = self.player_scores()
        print("\n\(self)")
        print(
            "Score: \(get_color(score.black, Disk.black.color())) |",
            "\(get_color(score.white, Disk.white.color()))"
        )
    }

    /// Check that the given coordinates are inside the board.
    private func check_coordinates(_ x: Int, _ y: Int) -> Bool {
        x >= 0 && x < self.size && y >= 0 && y < self.size
    }

    /// Count and return the number of black and white disks.
    private func player_scores() -> (black: Int, white: Int) {
        var black = 0
        var white = 0
        for disk in self.board {
            switch disk {
                case .white:
                    white += 1
                case .black:
                    black += 1
                case .empty:
                    break
            }
        }
        return (black, white)
    }

    /// Returns the total score.
    /// Positive value means more white disks and negative means more black disks.
    private func score() -> Int {
        self.board.reduce(0) { $0 + $1.rawValue }
    }

    /// Returns the state of the board (empty, white, black) at the given square.
    private func get_square(_ square: Square) -> Disk? {
        self.check_coordinates(square.x, square.y) ? self
            .board[square.y * self.size + square.x] : nil
    }

    /// Sets the given square to given value.
    private func set_square(_ square: Square, _ disk: Disk) {
        if !self.check_coordinates(square.x, square.y) {
            print("Invalid coordinates \(square)!")
        }
        self.board[square.y * self.size + square.x] = disk
    }
}

extension Board: CustomStringConvertible {
    var description: String {
        var text = " "
        for x in 0 ..< self.size {
            text += " \(x)".dim()
        }
        for y in 0 ..< self.size {
            text += "\n\(y)".dim()
            for x in 0 ..< self.size {
                let disk = self.board[y * self.size + x]
                text += " \(disk.board_char())"
            }
        }
        return text
    }
}
